### PR TITLE
[Docs] Rewrite 'Trees' section

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Rewrite 'Trees' section as part of the documentation week.

WIP by @Axel-Naumann, @jalopezg-r00t, and @etejedor.

### CODE REVIEW

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:140]: Maybe clarify a bit this comment.

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:182]:
```suggestion
if a member is a `std::vector<X>`, the tree can split `X` into sub-branches, too.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:238]: I think iteration over entries is performed this way.  @etejedor can you confirm?

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:247]: Maybe add a simple example?

**Review by jalopezg-r00t — COMMENTED** [review]
See comments.

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:30]: We will word that in an even stronger way!

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:238]: He confirms!

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:48]: > `contain data`

Might be an over simplifications.  Literally speaking the only entity that contains the data  are:
* the TBasket (in memory, uncompressed)
* the region of TFile corresponding to this TBasket (on disk, compressed)
* the user variable (in memory.

And it is always the TBranch that holds the information about the baskets.

The distinction between branches and leaves is "whom" hold the address of the memory corresponding to the user variable that hold the data or "whom" can give access to the user variable address and type information (`GetAddress`, `GetValue`).

so maybe `give access to the data` is more accurate.

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:64]: The "unique" part is misleading since there is shared owership between the user and the `TFile`.  A raw pointer is likely adequate here as *usually* the user should not have to worry about deleting the TTree (but they can if they want to).

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:83]:
```suggestion
> The objects *and* variables used to create branches must not be destroyed until the `TTree` is deleted or `TTree::ResetBranchAddress()` is called.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:293]: I would rephrase this a bit.  Maybe something along the lines of
```suggestion
> Care must be taken to ensure that the order of entries in the primary tree matches friends' entries. This is especially relevant when processing a tree in parallel to generate a friend tree, as the entries might be written out in an undefined order (misaligned entries).
```

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:116]:
```required_change
The default split level is 99 which means to split all members at any recursion level.
```

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:122]: Technically that is during on `TBranch::Fill`.  For example if we have (mostly likely indirectly rather than this improbable simplification :) ):
```
struct SomeClass {
   Other *fPointer;
   Ohter *fAnotherPointer; 
};
```
and `fPointer` and `fAnotherPointer` points to the same address **and** the struct `SomeClass` is stored with split level 1 or higher, then the pointee object will be stored twice (and more annoyingly when retrieving the object, `fPointer` and `fAnotherPointer` will not point to two distinct objects .... hence the need/recommendation to use references (like TRef) or indices instead of pointers.

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:158]:
```suggestion
When filling a {% include ref class="TTree" %} into a ROOT file and if the ROOT file size reaches the value stored in the [TTree::GetMaxTreeSize()](https://root.cern/doc/master/classTTree.html#aca38baf017a203ddb3119a9ab7283cd9){:target="_blank"}, the current ROOT file is closed and a new ROOT file is created. If the original ROOT file is named `myfile.root`, the subsequent ROOT files are named `myfile_1.root`, `myfile_2.root`, etc.
```

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:160]: We should consider adding text about `AutoFlush` and clusters around here.

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:208]:
```suggestion
You can select or deselect branches from being read by `TTree::GetEntry()` by calling [`TTree::SetBranchStatus()`](https://root.cern/doc/master/classTTree.html#aeca53bcd4482de1d883d259df6663920).
```

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:235]:
```suggestion
First you insert the tree entry numbers you want to process into the `TEntryList`.
```

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:262]:
```suggestion
it wants to know the name of the trees in the files, and the file names, and will act as if it was a huge, continuous tree:
```
> (has to be the same throughout),

TChain actually works properly with names change as long as it is specified when adding the file to the chain.

File: manual/trees/index.md
Comment (pcanal) [inline manual/trees/index.md:283]:
```suggestion
While updating an existing tree is non-trivial, extending it with additional branches, potentially an "improved" version of an original branch, is straighforward.
```

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:64]: I think the key is:

> but they can if they want to

That means we should use the best practices pattern, and that's a stack variable - which is awkward because of ROOT's memory management requirements, and thus the secon best option is a `unique_ptr`. Do you see a *problem* with that, or is it "just" misleading?

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:98]: Discusses with Enric, clarification can quickly go out of control here :-) and we agree this is "good enoough".

File: manual/trees/index.md
Comment (pcanal) [reply manual/trees/index.md:64]: it is just misleading.

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:48]: Done!

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:116]: Done!

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:122]: Thanks for the clarification! Addressed.

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:158]: We rephrased this, we think it makes more sense now.

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:141]: Done

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:222]: Applied

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:328]: Clarified

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md]: Addressed

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md]: Addressed (differently)

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:62]:
```suggestion
TTree finds the baskets for a given entry for a given branch by means of a header stored in the file.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:143]:
```suggestion
For the general case, indices into object collections could be persisted instead of pointers.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:222]:
```suggestion
By default this happens approximately every 30MB of compressed data.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:388]:
```suggestion
ROOT offers different ways to examine tree structure and its contents, from text to graphics.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:516]:
```suggestion
- Use the [TTree::BuildIndex()](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bd0b06eea6c12b998){:target="_blank"} method to build an index table using expressions depending on the value in the leaves.
```

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:526]: Broken link?

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:541]: Broken link?

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:546]:
```suggestion
The most convenient place to create the index is at the end of the filling process just before saving the tree header. If a previous index was available, it will be overridden by this new call.
```

**Review by jalopezg-r00t — COMMENTED** [review]
See comments.

File: manual/trees/index.md
Comment (etejedor) [inline manual/trees/index.md:523]: Says above but the example is below?

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:530]: https://root.cern.ch/doc/master/classTTree.html#a8c77a02a1323019375d273299a2c14b1

Comment (etejedor) [reply manual/trees/index.md:530]: Also, I believe we never include the parameters when we link to a method, so I would remove `majornumber, minornumber`

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:523]: Should we say they are leaves?

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:143]: Persistify is actually the transitive verb here, so we think it's ok.

File: manual/trees/index.md
Comment (etejedor) [inline manual/trees/index.md:63]: Is it ok to talk about `event range` when so far we've talked about `entries` ?

File: manual/trees/index.md
Comment (pcanal) [reply manual/trees/index.md]: No, they are really expressions (eg. `run%1000`, `event_part_1 * 1000 + event_part_2`)

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:76]:
```suggestion
It derives from {% include ref class="TTree" %} and is constructed with a list of column names separated by `:`.
```

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed thanks!

File: manual/trees/index.md
Comment (jalopezg-r00t) [inline manual/trees/index.md:400]: If this is going to be merged in its current state, please remove this!
```suggestion
```

**Review by jalopezg-r00t — APPROVED** [review]
As discussed on Friday, the documentation week ended and this should be merged.
It is in a much better state now, but see suggestion before merging.

File: manual/trees/index.md
Comment (etejedor) [reply manual/trees/index.md:400]: Nice catch thanks!

**Comment by etejedor** [comment]
Thank you @jalopezg-r00t for the catch! All comments have been resolved so I'll merge if checks are fine.

**Comment by etejedor** [comment]
Still can't merge, for two reasons:
- There is a link from the histograms section to an example that we removed in trees. To be discussed today in the meeting how to solve that (probably just removing the reference to the example and replacing it with a reference to the trees section).
- We have links to I/O from trees but I/O still doesn't exist. To be discussed how to proceed (I'd perhaps comment out those links for now to make it possible to merge trees).

**Review by pcanal — APPROVED** [review]

File: manual/trees/index.md
Comment (Axel-Naumann) [reply manual/trees/index.md:374]: Thanks, addressed.

**Pull request merged**
